server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -525,9 +533,15 @@ private void searchLeaf(LeafReaderContext ctx, Weight weight, Collector collecto
                         leafCollector,
                         this.cancellable.isEnabled() ? cancellable::checkCancelled : () -> {}
                     );
-                } catch (CollectionTerminatedException e) {
+                } catch (CollectionTerminatedException | TimeLimitingCollector.TimeExceededException e) {
+                    leafCollector.finish();
```

Comment:
I was looking at where finish is called in LUcene's IndexSearcher. Do we need to call it from within the catch? Don't we need to call if either way ? I may be getting confused.

Reply:
@javanna I am thinking we need to call `finish()` if any results are collected before we adjust the leaf collector as different collectors may reset different state via `finish()`.

Reply:
In lucene
```
 // Note: this is called if collection ran successfully, including the above special cases of
      // CollectionTerminatedException and TimeExceededException, but no other exception.
      leafCollector.finish();
```

I guess we could just add `leafCollector.finish()` to the end of the method and have our exception catching prevent the method from exiting before the finish is called.